### PR TITLE
Include units in csv column names

### LINKED ISSUE #1514: [OM] Include units in csv column names
This is part 5 of the issue #1416 

- [x] Include units directly in the csv column names.
- [x] Remove extra columns like `.prefix`, `.units`.

Currently, it is already possible to include units when using csv filtering.

For example, with this filter

```
Weather.average_annual_temperature
```

I got the following

```
Weather.average_annual_temperature,Weather.average_annual_temperature.units,Weather.average_annual_temperature.prefix
8.178051643192488,°C,Weather
```

### PR DESCRIPTION
**Pull request by ltp23**
### Context

Issue(s) closed by this pull request: closes #1514 

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.

### What & Why
When using csv filtering (`csv_<example>.txt`), there are additional columns such as `.prefix` and `.units` that seem redundant and use extra memory storage. This PR aims to remove those unnecessary columns and directly incorporate the units into the column names.

When the exclude-info-maps flag `-i` is used, that will disable this feature because information about units is stored in the info maps.

This feature doesn't apply to reports yet (which are also in the csv format) because there are some other considerations such as missing info maps and different combinations of aggregation operations affecting the final units. There is an open issue regarding this #1516.

### How

- [x] Look for the units key in the `info_maps` dictionary if it exists. Ignore other keys in the `info_maps`.
- [x] Format the units somehow. This PR only does a simple appending of the units string to the variable name `simple_variable_name_<units>` or `<dictionary_variable_name>.subkey_<units>`.
- [x] Do some code refactoring and write/update unit tests
- [x] There are some missing units in the animal module reporter and they need to be added.

### Test plan

To test things out, I would create a filter file `csv_examples.txt` and select simple and dictionary variables.

```
Weather.average_annual_temperature
^AnaerobicDigestion_ManureTreatmentDailyOutput_.*LAC_COW.liquid_manure_total_volatile_solids$
AnimalModuleReporter.report_ration_interval_data.ration_nutrient_amount_.*LAC_COW
```

### CODE REVIEW

**Review by ew3361zh — APPROVED** [review]
Nice work, Loi. I tested this with a few filters and then one of the currently nested units dicts and it accounted for them all well.

I wondered if adding the units at the end might be more readable if they were added in parantheses?

e.g.

`ManureTreatmentDailyOutput_Pen_1_GROWING.daily_final_manure_volume_m^3`
->
`ManureTreatmentDailyOutput_Pen_1_GROWING.daily_final_manure_volume_(m^3)`

Testing looked good as well.

File: RUFAS/output_manager.py
Comment (ew3361zh) [inline RUFAS/output_manager.py:584]: I think an example could be helpful here to understand what the subkey is.

Comment (ltp23) [reply RUFAS/output_manager.py:584]: I have added some examples in the docstring.

**Comment by ltp23** [comment]
> I wondered if adding the units at the end might be more readable if they were added in parantheses?
> 
> e.g.
> 
> `ManureTreatmentDailyOutput_Pen_1_GROWING.daily_final_manure_volume_m^3` -> `ManureTreatmentDailyOutput_Pen_1_GROWING.daily_final_manure_volume_(m^3)`

Thanks Niko. I have just made a change and put the units in parentheses.

File: RUFAS/output_manager.py
Comment (Isgotkowitz) [inline RUFAS/output_manager.py]:
```suggestion
        return f" ({units[subkey]})"
```
Instead of appending with an underscore, I think it would be cleaner to just the units enclosed in parentheses. This makes the column a little more readable, especially for non-devs who don't spend all their reading and writing things in snake case.

File: RUFAS/output_manager.py
Comment (Isgotkowitz) [inline RUFAS/output_manager.py]: If the units aren't present for a variable, I think that merits logging an error/or warning

Comment (Isgotkowitz) [inline RUFAS/output_manager.py]: If the subkey is not in the units, I think that should be its own case that causes an error to be logged. And likewise for the case when the subkey is "None".

File: RUFAS/output_manager.py
Comment (Isgotkowitz) [inline RUFAS/output_manager.py:554]: This leaves the door open for a pretty serious issue: what if the same variable is recorded with varying units? It seems unlikely that this would ever happen, but currently it is possible and we should consider whether or not it should be allowed. And if it is decided that it shouldn't be allowed then it is an opportunity for some large reductions in the memory footprint of info maps, because the units would only have to be recorded once and not in every single info map.

Note that this issue is well outside the scope of this PR, I just want to bring it up so that we can start considering solutions.

**Review by Isgotkowitz — COMMENTED** [review]
Looks good Loi! Wrote some filters and they were behaving as expected. I left some suggestions and thoughts, but they are take it or leave it.

I am holding off on approving this just because I think the second approval should come from a SME, and this PR already has one dev review

File: RUFAS/output_manager.py
Comment (ltp23) [reply RUFAS/output_manager.py]: Okay. I will add a whitespace between the variable name and the units.

File: RUFAS/output_manager.py
Comment (ltp23) [reply RUFAS/output_manager.py]: Both CSV filtering and report generation share the same method of making CSV files. Currently, at the point of making CSVs from reports, the info maps have been stripped so this logging would cause lots of warnings. I haven't yet worked out how to incorporate units into the reports gracefully considering the existing usage by pilot testers.

Comment (ltp23) [reply RUFAS/output_manager.py]: I can rearrange the conditions a little bit to make things clearer.

**Review by jadamchick — COMMENTED** [review]
Hi Loi, I like it! 
For some reason it didn't seem to work for the first set of filters I tried (relevant to manure excretion outputs) -- though it looks like those units exist in the code (and when I run the same CSV on main). 
```
AnimalModuleReporter.report_animal_module_manure.daily_aggregate_.*_total_ammoniacal_nitrogen_concentration
AnimalModuleReporter.report_animal_module_manure.daily_aggregate_.*_urine_nitrogen
AnimalModuleReporter.report_animal_module_manure.daily_aggregate_.*_manure_nitrogen
AnimalModuleReporter.report_animal_module_manure.daily_aggregate_.*_phosphorus
AnimalModuleReporter.report_animal_module_manure.daily_aggregate_.*_phosphorus_fraction
AnimalModuleReporter.report_animal_module_manure.daily_aggregate_.*_total_solids
AnimalModuleReporter.report_animal_module_manure.daily_aggregate_.*_degradable_volatile_solids
AnimalModuleReporter.report_animal_module_manure.daily_aggregate_.*_non_degradable_volatile_solids
```

but when I tried with a different set, it seemed to work fine (including the cull reason units which are also within a dictionary). 

```
^AnimalModuleReporter.report_life_cycle_manager_data.*
^AnimalModuleReporter.report_daily_animal_population.*
```

Any thoughts on why the units didn't come through on the set for `report_animal_module_manure`?

Let me know if that is something you want to address and get a re-review before approval, or if you think it is something unrelated to this PR.

File: RUFAS/output_manager.py
Comment (ltp23) [reply RUFAS/output_manager.py:554]: Because I'm using the hardcoded index 0, I'm assuming that the units should stay consistent for all the add-variables calls to the same variable. But this can change if we want to accommodate the case where units can be dynamic. 

I also think we could perform a check when we add a variable to the OM pool to make sure the current units are the same as the last added units.

**Comment by ltp23** [comment]
Hi Julie (@jadamchick), there is a small error when reporting units for several variables but I have fixed it now.

```
AnimalModuleReporter.report_animal_module_manure.daily_aggregate_.*
AnimalModuleReporter.report_pen_manure_properties.pen_.*
SimulationEngine.simulate.available_feeds_on_final_day
```

or use the catch-all filter

```
.*
```

Please let me know if it also works on your end.

**Review by jadamchick — APPROVED** [review]
Works great now!

**Pull request merged**
